Add full IP pinning enforcement

AntonAM opened the pull request:
We're adding IP pinning check to `authorizer.Authorize` which is used for every call, so now all communications with teleport should enforce IP pinning. Also making sure we always provide login IP for user certificate creation, with single use certificates always pinned (if it's enterprise), and correct client IP propagation everywhere.

AntonAM commented on api/client/proxy.go:
```diff
@@ -65,15 +70,24 @@ func DialProxyWithDialer(
 
+	var signedHeader []byte
+	var err error
+	if cfg.proxyHeaderGetter != nil {
```
Yes, I actually had this in mind to simplify with the new dialer I added, that's why I was getting rid of `*net.Dialer` in parameters everywhere. But then got derailed and didn't return to it. Now done 👍

AntonAM commented on api/client/proxy.go:
```diff
@@ -88,22 +102,55 @@ func dialProxyWithHTTPDialer(
+		}
+
+		// matching the behavior of tls.Dial
+		cfg := tlsConfig
```
Yep, I was thinking about copying `TLSDial` function from utils package. Done (and used in another place we had similar need).

AntonAM commented on lib/auth/grpcserver.go:
```diff
@@ -2613,14 +2614,19 @@ func userSingleUseCertsGenerate(ctx context.Context, actx *grpcContext, req prot
+	}
+
+	if modules.GetModules().BuildType() == modules.BuildEnterprise {
+		opts = append(opts, certRequestPinIP(true)) // We always want to pin single use certs to client's IP, regardless of roles)
```
That is mentioned in the MFA RFD 14 ( https://github.com/gravitational/teleport/blob/master/rfd/0014-session-2FA.md#constraints ) and @strideynet pointed it out in discussions about IP pinning.

AntonAM commented on lib/srv/alpnproxy/proxy.go:
```diff
@@ -409,6 +418,30 @@ func (p *Proxy) handleConn(ctx context.Context, clientConn net.Conn, defaultOver
+		return trace.Wrap(err)
+	}
+
+	if identity.PinnedIP != "" && clientIP != identity.PinnedIP {
```
We do a bit different thing here, we don't check user's roles, only checking if pinned IP on the cert doesn't match observed one.

AntonAM commented on lib/authz/permissions.go:
```diff
@@ -289,6 +295,40 @@ func (a *authorizer) fromUser(ctx context.Context, userI interface{}) (*Context,
+	}
+
+	if clientIP != identity.PinnedIP {
+		logrus.WithFields(logrus.Fields{
```
Hm, I think this is perfect place for this debug log, there's one direct caller - authorizer and myriad of higher level callers that don't do logging. And they actually don't really have direct access to the ip fields which we log here.

tigrato replied:
This is a problem because you are using the default logger which is not configured with all the configurations provided in the config.

I still think the caller must handle the log here since it has access to the configured logger.

You can `return trace.Wrap(ErrIPPinningMismatch, "pinned IP and client IP mismatch: client_ip: %v, pinned_ip: %v", clientIP,identity.PinnedIP)`

With this, you still have errors.Is working and you can log it with the correct logger.

AntonAM replied:
I'm hesitant to return specific IPs in error for security concerns. But I agree that using default logger is not ideal. In this case main problem I think that we're missing component field. I've changed `CheckIPPinning` to accept logger and currently authorizer will just provide logger with `trace.Component: "authorizer"` and I will handle adding logger to `NewAuthorizer` in a separate PR (issue for this https://github.com/gravitational/teleport/issues/25035 )

AntonAM commented on lib/authz/permissions.go:
```diff
@@ -1025,6 +1065,9 @@ func ConvertAuthorizerError(ctx context.Context, log logrus.FieldLogger, err err
 		// user not found, wrap error with access denied
 		return trace.Wrap(err, "access denied")
+	case errors.Is(err, ErrIPPinningMissing) || errors.Is(err, ErrIPPinningMismatch):
+		log.Warn(err)
```
I think this is fine, we do already log errors in this function, in default case and access denied case, so callers already rely on it being logged here, so I think this is consistent.

AntonAM commented on lib/multiplexer/multiplexer.go:
```diff
@@ -686,13 +686,14 @@ func NewPROXYSigner(signingCert *x509.Certificate, jwtSigner JWTPROXYSigner) (*P
-			"dst_addr":     fmt.Sprintf("%v", destination),
-			"cluster_name": p.clusterName}).Warn("Got bad IP while trying to sign PROXY header")
+		logger.Warn("Got bad IP while trying to sign PROXY header")
 		return nil, nil
```
There was a few cases when we had situation that was not really relevant to supported parts of IP pining but failed IP propagation. So it was a bit dangerous place which could affect all users instead of just pinned ones. I think we probably can remove it now, since all cases should be theoretically handled, at least for the testplan duration 🤔

AntonAM commented on lib/kube/proxy/forwarder.go:
You mean direct dialer in the kube forwarder? Yep, I will, but in the separate PR to not complicate this one.

probakowski commented on api/client/client.go:
```diff
@@ -485,6 +485,16 @@ func (c *Client) grpcDialer() func(ctx context.Context, addr string) (net.Conn,
+			if err != nil {
+				return nil, trace.ConnectionProblem(err, "unable to get signed PROXY header: %v", err)
+			}
+			_, err = conn.Write(signedHeader)
+			if err != nil {
+				return nil, trace.ConnectionProblem(err, "could not write signed PROXY header into connection: %v", err)
+			}
```
```suggestion
			if _, err := conn.Write(signedHeader); err != nil {
				return nil, trace.ConnectionProblem(err, "could not write signed PROXY header into connection: %v", err)
			}
```

probakowski commented on api/client/contextdialer.go:
```diff
@@ -96,12 +107,37 @@ func newDirectDialer(keepAlivePeriod, dialTimeout time.Duration) *net.Dialer {
+			return nil, trace.Wrap(err)
+		}
+
+		if headerGetter != nil {
+			signedHeader, err := headerGetter()
+			if err != nil {
+				conn.Close()
+				return nil, trace.Wrap(err)
+			}
+			_, err = conn.Write(signedHeader)
+			if err != nil {
+				conn.Close()
+				return nil, trace.Wrap(err)
+			}
+		}
```
We have very similar code in `client.go`, maybe extract helper method?

AntonAM replied:
Yeah, I refactored it code in `client.go` to make use of proxy header dialer as well, so now we only have it one place.

probakowski commented on api/utils/tlsutils/tlsutils.go:
```diff
@@ -36,3 +40,67 @@ func ParseCertificatePEM(bytes []byte) (*x509.Certificate, error) {
+	// from the hostname we're connecting to.
+	if tlsConfig.ServerName == "" {
+		// Make a copy to avoid polluting argument or default.
+		c := tlsConfig.Clone()
+		c.ServerName = hostname
+		tlsConfig = c
```
```suggestion
		tlsConfig = tlsConfig.Clone()
		tlsConfig.ServerName = hostname
```

probakowski commented on api/utils/tlsutils/tlsutils.go:
```diff
@@ -36,3 +40,67 @@ func ParseCertificatePEM(bytes []byte) (*x509.Certificate, error) {
+
+	conn := tls.Client(plainConn, tlsConfig)
+	errC := make(chan error, 1)
+	go func() {
+		err := conn.HandshakeContext(ctx)
+		errC <- err
+	}()
+
+	select {
+	case err := <-errC:
+		if err != nil {
+			plainConn.Close()
+			return nil, trace.Wrap(err)
+		}
+	case <-ctx.Done():
+		plainConn.Close()
+		return nil, trace.BadParameter("tls handshake has been canceled due to timeout")
+	}
```
Do we need this? Won't `conn.HandshakeContext` fail by itself if context times out?

AntonAM replied:
Hm,  I just copied it over from `tlsdial.go`, but yep, seems to be redundant 🤔 Removing.

probakowski commented on lib/client/api.go:
```diff
@@ -3020,8 +3024,23 @@ func makeProxySSHClient(ctx context.Context, tc *TeleportClient, sshConfig *ssh.
+// CreatePROXYHeaderGetter returns PROXY headers signer with embedded client source/destination IP addresses,
+// which are taken from the context.
+func CreatePROXYHeaderGetter(ctx context.Context, proxySigner multiplexer.PROXYHeaderSigner) client.PROXYHeaderGetter {
+	if proxySigner != nil {
+		src, dst := utils.ClientAddrFromContext(ctx)
+		if src != nil && dst != nil {
+			return func() ([]byte, error) {
+				return proxySigner.SignPROXYHeader(src, dst)
+			}
+		}
+	}
```
We could invert some of the conditions to reduce nesting. Like
```go
if proxySigner == nil {
    return nil
}
```

AntonAM replied:
Done.

probakowski reviewed (commented):
Overall looks ok, few suggestions

AntonAM committed:
Reduce nesting

AntonAM committed:
Use proxyHeader dialer in authConnect

AntonAM committed:
Add full IP pinning enforcement

We're adding IP pinning check to `authorizer.Authorize` which is used for every call,
so now all communications with teleport should enforce IP pinning.
Also making sure we always provide login IP for user certificate creation
and correct client IP propagation everywhere.

AntonAM committed:
Remove redundant channels processing in TLSDial

AntonAM committed:
Clone TLS config

AntonAM committed:
Wider use proxyHeaderDialer and remove adhoc writing of singed header

AntonAM committed:
Wrap error

Co-authored-by: Tiago Silva <tiago.silva@goteleport.com>

AntonAM committed:
Add godocs

AntonAM committed:
Fix wording

Co-authored-by: Tiago Silva <tiago.silva@goteleport.com>

AntonAM committed:
Improve proxyHeaderSigner usage

AntonAM committed:
Simplify tlsConfig manipulation

Co-authored-by: Przemko Robakowski <przemko.robakowski@goteleport.com>

AntonAM committed:
Add integration test for App IP pinning.

AntonAM committed:
Add helper function TLSDial

AntonAM committed:
Update generated protobufs

AntonAM committed:
Remove ignoring of bad IP on signed PROXY header generation

tigrato commented on lib/multiplexer/multiplexer.go:
```diff
@@ -686,13 +686,9 @@ func NewPROXYSigner(signingCert *x509.Certificate, jwtSigner JWTPROXYSigner) (*P
-		return nil, nil
-	}
-
+	log.WithFields(log.Fields{
```
Is this correct? If err!=nil it isn't successful, isn't it?

AntonAM replied:
Nice catch, thanks, fixed it 👍

AntonAM committed:
Provide logger to CheckIPPinning function
